### ISSUE TITLE
Attachment Picker Doesn't Work/Crashes App

### ISSUE THREAD

**Issue by jstheoriginal** [zendesk/zendesk_sdk_chat_ios#93]
I tried to choose an image from my Photo Library and this comes up and no picker appears and it says something about trying to present the UIImagePickerController on UIViewController who's view is not in the window hierarchy.  I then chose Camera instead, and it crashes. Log below.

```
2016-06-02 17:12:31.556 [22151:282622] Warning: Attempt to present <UIImagePickerController: 0x7bc4f400> on <UIViewController: 0x7a654810> whose view is not in the window hierarchy!
2016-06-02 17:13:55.997 [22151:282622] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Source type 1 not available'
*** First throw call stack:
(
	0   CoreFoundation                      0x05d68494 __exceptionPreprocess + 180
	1   libobjc.A.dylib                     0x05822e02 objc_exception_throw + 50
	2   UIKit                               0x04271a2b -[UIImagePickerController sourceType] + 0
	3   SkipTheDishes                       0x00abc04e -[ZDCChatView actionSheet:didDismissWithButtonIndex:] + 160
	4   UIKit                               0x04674753 __78-[UIActionSheet _performPresentationDismissalWithClickedButtonIndex:animated:]_block_invoke + 85
	5   UIKit                               0x0498510d __78-[_UIAlertControllerShimPresenter _dismissAlertControllerAnimated:completion:]_block_invoke + 140
	6   UIKit                               0x04985008 -[_UIAlertControllerShimPresenter _dismissAlertControllerAnimated:completion:] + 551
	7   UIKit                               0x046746cb -[UIActionSheet _performPresentationDismissalWithClickedButtonIndex:animated:] + 176
	8   UIKit                               0x04674d11 -[UIActionSheet _dismissForTappedIndex:] + 48
	9   UIKit                               0x04674b32 __37-[UIActionSheet _prepareAlertActions]_block_invoke + 60
	10  UIKit                               0x041f84d1 -[UIAlertController _fireOffActionOnTargetIfValidForAction:] + 103
	11  UIKit                               0x041f8d60 __85-[UIAlertController _dismissAnimated:triggeringAction:triggeredByPopoverDimmingView:]_block_invoke + 46
	12  UIKit                               0x04001db6 -[UIPresentationController transitionDidFinish:] + 1521
	13  UIKit                               0x040059ee __56-[UIPresentationController runTransitionForCurrentState]_block_invoke_2 + 224
	14  UIKit                               0x049dbc5f -[_UIViewControllerTransitionContext completeTransition:] + 118
	15  UIKit                               0x0448b5c2 __61-[_UIAlertControllerAnimatedTransitioning animateTransition:]_block_invoke61 + 133
	16  UIKit                               0x04a941d2 __171-[UIAlertControllerVisualStyleActionSheet animateAlertControllerView:ofAlertController:forPresentation:inContainerView:descendantOfContainerView:duration:completionBlock:]_block_invoke81 + 26
	17  UIKit                               0x03f2bd70 -[UIViewAnimationBlockDelegate _didEndBlockAnimation:finished:context:] + 647
	18  UIKit                               0x03f0508f -[UIViewAnimationState sendDelegateAnimationDidStop:finished:] + 247
	19  UIKit                               0x03f05485 -[UIViewAnimationState animationDidStop:finished:] + 90
	20  QuartzCore                          0x039e168b _ZN2CA5Layer23run_animation_callbacksEPv + 305
	21  libdispatch.dylib                   0x067269cd _dispatch_client_callout + 14
	22  libdispatch.dylib                   0x0670bf7c _dispatch_main_queue_callback_4CF + 910
	23  CoreFoundation                      0x05cb21be __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 14
	24  CoreFoundation                      0x05c70434 __CFRunLoopRun + 2356
	25  CoreFoundation                      0x05c6f846 CFRunLoopRunSpecific + 470
	26  CoreFoundation                      0x05c6f65b CFRunLoopRunInMode + 123
	27  GraphicsServices                    0x07e99664 GSEventRunModal + 192
	28  GraphicsServices                    0x07e994a1 GSEventRun + 104
	29  UIKit                               0x03e66eb9 UIApplicationMain + 160
	30  SkipTheDishes                       0x00168c01 main + 145
	31  libdyld.dylib                       0x06750a25 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

Obviously the simulator shouldn't actually work since there is no camera, but it shouldn't crash...this should be handled and not do anything after checking if the camera is available. Normally you need to do some checking to make sure that the Photo Library and/or Camera are available.  What if the user didn't allow permission to access the camera or photos? Would it crash then?

**Comment by tecknut** [comment]
Can you provide the context around the image picker warning?

```
2016-06-02 17:12:31.556 [22151:282622] Warning: Attempt to present <UIImagePickerController: 0x7bc4f400> on <UIViewController: 0x7a654810> whose view is not in the window hierarchy!
```

We've not seen this before, it would be good to track this down.

**Comment by jstheoriginal** [comment]
Sure. 

In the chat, I tapped the attach button. The photo library and camera options appear. I tap the photo library button. The error shows up in console and nothing pops up on screen.

**Comment by oarrabi** [comment]
Hello @jstheoriginal,

I added an item on our bug tracker to fix the issue of showing the camera/library buttons in the action sheet without checking.
I however cant reproduce the issue that you are experiencing.
Would it be possible to share a sample project that reproduces it?
Alternatively, can you share a snippet on how you show the chat sdk? and more info about your view controller/windows heirarchy?

Thanks,

**Comment by jstheoriginal** [comment]
Ok @oarrabi ! This will hopefully be able to solve a lot of issues. I've uploaded a sample project here: https://www.dropbox.com/s/a8zxwx0n1dx8bix/CustomTabBarZendeskIssue.zip?dl=0 (please let me know once you've grabbed it so I can delete it). You'll need to put in a valid Zendesk account key in the App Delegate.

The blue view is where we have our tab bar in our app.

I tested the sample project and it behaves the exact same as our app does.  On device, it doesn't crash when you choose the Photo Library or Camera for an attachment, but nothing appears when you tap the buttons.  The warning appears in the debugger ```Warning: Attempt to present <UIImagePickerController: 0x13515f600> on <UIViewController: 0x134e630f0> whose view is not in the window hierarchy!```

This project should also be used to fix the issues with the blank space above the keyboard.  

Ticket about the space above the keyboard in pre-chat and chat when keyboard showing: https://github.com/zendesk/zendesk_sdk_chat_ios/issues/89
Ticket about textview issues could be linked to the constraints breaking: https://github.com/zendesk/zendesk_sdk_chat_ios/issues/87

Here are some screenshots I took, and some Reveal snapshots to see what is happening with the window hierarchy.

![img_0288](https://cloud.githubusercontent.com/assets/5117473/16052373/949ee6de-3229-11e6-99c4-bd0759a5200b.jpg)

![img_0289](https://cloud.githubusercontent.com/assets/5117473/16052374/94a27682-3229-11e6-8860-1665e7d1ecae.jpg)

![img_0290](https://cloud.githubusercontent.com/assets/5117473/16052375/94a7c722-3229-11e6-8a2c-ac200573022a.jpg)

![screenshot 2016-06-14 11 56 55](https://cloud.githubusercontent.com/assets/5117473/16052390/b51fad08-3229-11e6-90cf-dd5ae6014f92.png)

![screenshot 2016-06-14 11 59 14](https://cloud.githubusercontent.com/assets/5117473/16052391/b53b7916-3229-11e6-95ef-45c69eefe1a5.png)

In Reveal, it looks like it's showing the ZDCChatViewController when the keyboard is presented (which is the first Reveal screenshot above).  When the keyboard isn't presented, it looks like the UITableViewWrapperView is behind the custom tab bar....

**Comment by oarrabi** [comment]
Hello @jsthteoriginal,

I am super sorry for not updating this issue for a while.
Thanks to the demo we were able to reproduce and finally fix this issue and https://github.com/zendesk/zendesk_sdk_chat_ios/issues/89. Thanks so much on providing it, it really made a difference.

Both the issues are fixed and will be released in the upcoming release (which should be published within the next 2 weeks).

Thanks for your patience :)

**Comment by jstheoriginal** [comment]
@oarrabi yay! glad it helped and glad it'll be fixed in the next release. 👍

**Comment by ndobri** [comment]
@jstheoriginal this is fixed with [Chat SDK 1.3](https://github.com/zendesk/zendesk_sdk_chat_ios/releases/tag/1.3.0.1). Feel free to reopen if otherwise.
